### PR TITLE
feat: support to send logs

### PR DESCRIPTION
**Pull request by kuisathaverat**
- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/BuildInfo.java [hunk 1]
```diff
@@ -5,38 +5,70 @@
...
 import io.opentelemetry.api.common.Attributes;
 import io.opentelemetry.context.Context;
+import org.yaml.snakeyaml.nodes.CollectionNode;
 
+import javax.annotation.Nonnull;
+import java.util.Collections;
 import java.util.Map;
 
-final class BuildInfo {
+public final class BuildInfo {
```
Comment (cyrille-leclerc) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/BuildInfo.java:18]: In Otel code, they have the guiding principle to keep as much things as possible package protected and final. I like this. It's why I used package protected. Let's see what we want to do.

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/BuildInfo.java [hunk 2]
```diff
@@ -5,38 +5,70 @@
 
 package io.jenkins.plugins.opentelemetry.job.log;
 
+import com.google.common.base.Objects;
 import io.jenkins.plugins.opentelemetry.semconv.JenkinsOtelSemanticAttributes;
 import io.opentelemetry.api.common.Attributes;
 import io.opentelemetry.context.Context;
+import org.yaml.snakeyaml.nodes.CollectionNode;
```
Comment (cyrille-leclerc) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/BuildInfo.java:12]: useful? I can't find references in the code

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderBuildListener.java [hunk 3]
```diff
@@ -49,7 +39,7 @@ public OtelLogSenderBuildListener(BuildInfo buildInfo, Map<String, String> conte
     public synchronized final PrintStream getLogger() {
         if (logger == null) {
             try {
-                logger = new PrintStream(new PlainTextConsoleOutputStream(new OtelLogOutputStream(buildInfo, context)), false, "UTF-8");
```
Comment (cyrille-leclerc) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderBuildListener.java:52]: Why did you remove `PlainTextConsoleOutputStream`? To favor readability in Jenkins screens? what about the readability in the Otel Log storage?

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderFlowNode.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import jenkins.util.JenkinsJVM;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+import java.io.IOException;
+
+/**
+ * Sender for logs retaled to a flow node.
+ * See https://github.com/jenkinsci/pipeline-cloudwatch-logs-plugin/blob/master/src/main/java/io/jenkins/plugins/pipeline_cloudwatch_logs/CloudWatchSender.java#L108
+ */
+final class OtelLogSenderFlowNode extends OtelLogSender {
```
Comment (cyrille-leclerc) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderFlowNode.java:12]: Do we need today to differentiate OtelLogSenderFlowNode from OtelLogSenderRoot?

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        } catch (Exception x) {
+            return new BrokenLogStorage(x).stepLog(flowNode, complete);
+        }
+    }
+
+    @Nonnull
+    @SuppressFBWarnings(value = "BC_UNCONFIRMED_CAST", justification = "forBuild only accepts Run")
+    @Deprecated
+    @Override
+    public File getLogFile(@Nonnull FlowExecutionOwner.Executable build, boolean complete) {
```
Comment (cyrille-leclerc) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java:71]: Can we explain in javadoc why we override this deprecated method?

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/es/Retriever.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+import java.time.format.DateTimeFormatter;
+import java.util.TimeZone;
+
+import static org.elasticsearch.index.query.QueryBuilders.boolQuery;
+import static org.elasticsearch.index.query.QueryBuilders.matchQuery;
+
+/**
+ * This class retrieve logs from Elasticsearch.
+ */
+public class Retriever {
```
Comment (cyrille-leclerc) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/es/Retriever.java:38]: I would be happy to call it `ElasticsearchRetriever` rather than `Retriever`

**Review by cyrille-leclerc — COMMENTED** [review]
Super promising, THANKS

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/BuildInfo.java (hunk 2 above)
Comment (kuisathaverat) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/log/BuildInfo.java:12]: IO have to clean imports

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderBuildListener.java (hunk 3 above)
Comment (kuisathaverat) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderBuildListener.java:52]: if we do not do it we can not restore the Pipeline headers on the retrieve, we do not want the Jenkins UI lost functionality if you use the plugin.

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderFlowNode.java (hunk 4 above)
Comment (kuisathaverat) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderFlowNode.java:12]: no, I have to delete the two classes after merge your changes

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderBuildListener.java (hunk 3 above)
Comment (cyrille-leclerc) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderBuildListener.java:52]: can we have good readability in the observability backend if we keep Jenkins annotations?

**Comment by cyrille-leclerc** [comment]
@kuisathaverat I see benefits in merging asap in the feature branch, Do you see problems with that approach?

**Comment by kuisathaverat** [comment]
let me finish the tests and we can merge it.

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderBuildListener.java (hunk 3 above)
Comment (kuisathaverat) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderBuildListener.java:52]: we retrieve the log files in pages, we have to make some stress tests to find the limits of the retrieve from Jenkins, I am concerned about retrieving the full text console output on jobs with a  huge log (GB)

**Comment by kuisathaverat** [comment]
I am checking if I do not break anything, so if it pass the CI would be good to merge.

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/BuildInfo.java [hunk 7]
```diff
@@ -5,38 +5,98 @@
...
+        return getContextKey(TRACE_ID);
+    }
+
+    @CheckForNull
+    public String getSpanId() {
+        return getContextKey(SPAN_ID);
+    }
+
+    @Nullable
+    private String getContextKey(@Nonnull String key) {
```
Comment (cyrille-leclerc) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/BuildInfo.java:89]:
```suggestion
    private String getContextValue(@Nonnull String key) {
```
@kuisathaverat shouldn't it return the value for the given `key`?
